File: Cpp/322.零钱兑换.cpp
/*
 * @lc app=leetcode.cn id=322 lang=cpp
 *
 * [322] 零钱兑换
 */

#include <iostream>
#include <vector>

using namespace std;

// @lc code=start
class Solution {
private:
    int getCoins(vector<int>& faces, int dp[], int n) {
        if (n < 0) return -1;
        if (n == 0) return 0;
        if (dp[n] == 0) {
            int coin = -1;
            for (auto face : faces) {
                int val = getCoins(faces, dp, n - face);
                if (val >= 0) {
                    if (coin != -1) coin = min(coin, val + 1);
                    else coin = val + 1;
                }
            }
            dp[n] = coin;
        }
        return dp[n];
    }
public:
    int coinChange(vector<int>& coins, int amount) {
        int dp[amount + 1];
        memset(dp, 0, (amount + 1) * sizeof(int));
        for (int i = 0; i < coins.size(); i++) {
            if (coins[i] == amount) return 1;
            if (coins[i] < amount) dp[coins[i]] = 1;
        }
        
        return getCoins(coins, dp, amount);
    }
};
// @lc code=end


int main(int argc, const char * argv[]) {
    // insert code here...

    vector<int> coins;
    
    coins = {1,2,5};
    // 3
    cout << Solution().coinChange(coins, 11) << endl;

    coins = {2};
    // -1
    cout << Solution().coinChange(coins, 3) << endl;
    
    coins = {1};
    // 0
    cout << Solution().coinChange(coins, 0) << endl;
    
    coins = {1};
    // 1
    cout << Solution().coinChange(coins, 1) << endl;
    
    coins = {1};
    // 2
    cout << Solution().coinChange(coins, 2) << endl;

    return 0;
}

File: Swift/322.零钱兑换.swift
/*
 * @lc app=leetcode.cn id=322 lang=swift
 *
 * [322] 零钱兑换
 */

// @lc code=start
class Solution {
//    private func getCoins(_ coins: [Int],  dp: inout [Int], n: Int) -> Int {
//        if n < 0 { return -1 }
//        if n == 0 { return 0 }
//        if dp[n] == 0 {
//            var num = -1
//            for coin in coins {
//                let num_ = getCoins(coins, dp: &dp, n: n - coin)
//                if num_ >= 0 {
//                    if num == -1 { num = num_ + 1}
//                    else { num = min(max(num, 0), num_ + 1) }
//                }
//            }
//            dp[n] = num
//        }
//        return dp[n]
//    }
//    
//    func coinChange1(_ coins: [Int], _ amount: Int) -> Int {
//        if amount == 0 { return 0 }
//        var dp: [Int] = .init(repeating: 0, count: amount + 1)
//        
//        for coin in coins {
//            if (coin <= amount) { dp[coin] = 1 }
//        }
//        
//        return getCoins(coins, dp: &dp, n: amount)
//    }
    
    func coinChange(_ coins: [Int], _ amount: Int) -> Int {
        var dp: [Int] = .init(repeating: Int.max, count: amount + 1)
        dp[0] = 0
        
        for coin in coins {
            var i = 0
            while (i + coin) <= amount {
                if dp[i] != Int.max {
                    dp[i + coin] = min(dp[i] + 1, dp[i + coin])
                }
                i += 1
            }
        }
        
        return dp[amount] == Int.max ? -1 : dp[amount]
    }
}
// @lc code=end

func main() {
    var coins: [Int]
    var amount: Int
    var res: Int
    
    /**
     输入：coins = [1, 2, 5], amount = 11
     输出：3
     解释：11 = 5 + 5 + 1
     */
    coins = [1,2,5]
    amount = 11
    res = 3
    assert(res == Solution().coinChange(coins, amount))
    
    /**
     输入：coins = [2], amount = 3
     输出：-1
     */
    coins = [2]
    amount = 3
    res = -1
    assert(res == Solution().coinChange(coins, amount))
    
    /**
     输入：coins = [1], amount = 0
     输出：0
     */
    coins = [1]
    amount = 0
    res = 0
    assert(res == Solution().coinChange(coins, amount))
    
    /**
     输入：coins = [1], amount = 1
     输出：1
     */
    coins = [1]
    amount = 1
    res = 1
    assert(res == Solution().coinChange(coins, amount))
    
    /**
     输入：coins = [1], amount = 2
     输出：2
     */
    coins = [1]
    amount = 2
    res = 2
    assert(res == Solution().coinChange(coins, amount))
}
